src/dataflow/src/render/join.rs

```diff
@@ -294,4 +294,81 @@ where
             panic!("render_join called on invalid expression.")
         }
     }
+
+    /// Renders a join of an arrangement to a constant collection of size 1.
+    /// `input` should be of the form
+    /// `Join { inputs: [ArrangeBy{...}], implementation: Semijoin, ...}`
+    pub fn render_semi_join(&mut self, input: &RelationExpr, scope: &mut G, worker_index: usize) {
+        if let RelationExpr::Join {
+            inputs,
+            equivalences,
+            implementation: expr::JoinImplementation::Semijoin,
+            ..
+        } = input
+        {
+            if let RelationExpr::ArrangeBy {
+                input: inner_input,
+                keys,
+            } = &inputs[0]
```

Comment:
Afaict, if this test does not pass, we do nothing. Probably we should explicitly complain? Surely it doesn't work correctly on any other input?